csp_billing_adapter_microsoft/plugin.py

```diff
@@ -142,18 +234,30 @@ def _fetch_metadata(url):
         method='GET'
     )
     try:
-        value = urllib.request.urlopen(data_request).read()
-        return value.decode()
+        with urllib.request.urlopen(data_request) as value:
+            return value.read().decode("utf-8")
     except urllib.error.URLError as error:
-        log.error('Failed to retrieve metadata for: {url}. {error}'.format(
-            url=url,
-            error=str(error)
-        ))
-        return {}
+        log.error('Failed to retrieve metadata for: %s: %s', url, str(error))
+        return "{}"
 
 
-def _get_api_token():
-    """
-    Get the token to authenticate when using the Billing API
-   """
-    return json.loads(_fetch_metadata(TOKEN_URL))
+def _get_msi_token():
+    """Get the MSI token to authenticate when using the Billing API"""
+    # https://learn.microsoft.com/en-us/partner-center/marketplace/marketplace-metering-service-authentication
+
+    resource = '20e940b3-4c77-4b0b-9a53-9e16a1b010a7'
```

Comment:
If this needs to be a specified hardcoded value, probably worth putting a comment here to call that out, even if it is in the above documentation link.

Reply:
i will add additional text

Reply:
We have to double check with Msft that this magic ID stays the same for the packaged app.

Reply:
please review and see if the next text is sufficient

Reply:
I tested using their packaged app example and it works. I received this reply from Microsoft when I asked about the setting.  @rjschwei Does this address your concern?

```
Hi Keith,

 Regarding the “resource”, it needs to be “20e940b3-4c77-4b0b-9a53-9e16a1b010a7” as per [Metering API documentation](https://learn.microsoft.com/en-us/partner-center/marketplace/marketplace-metering-service-authentication#:~:text=Resource,9a53-9e16a1b010a7.). This is the audience(Metering API) for which the token needs to be generated here. 

Thanks and Regards,

Shashi Shekhar
```

Reply:
Comment looks good to me, happy to see this resolved.

Reply:
Could be an envvar set in helm chart. But only if we think it may change in the future.

Reply:
We can revisit this in the future if there is a need to paremitize it.